web/client-api/src/main/java/io/deephaven/web/client/api/barrage/stream/HandshakeStreamFactory.java

```diff
@@ -0,0 +1,138 @@
+                    };
+                },
+                (first, metadata) -> {
+                    Map<String, JsArray<Function>> listeners = listenerMap();
+                    io.deephaven.javascript.proto.dhinternal.grpcweb.grpc.InvokeRpcOptions<HandshakeRequest, HandshakeResponse> props =
+                            Js.cast(InvokeRpcOptions.create());
+                    props.setRequest(first);
+                    props.setHost(connection.browserFlightServiceClient().serviceHost);
+                    props.setMetadata(metadata);
+                    props.setTransport(null);// ts doesnt expose these two, stick with defaults for now
+                    props.setDebug(false);
+                    props.setOnMessage(responseMessage -> {
+                        listeners.get("data").forEach((item, index, arr) -> item.call(null, responseMessage));
+                    });
+                    props.setOnEnd((status, statusMessage, trailers) -> {
+                        listeners.get("status").forEach(
+                                (item, index, arr) -> item.call(null,
+                                        ResponseStreamWrapper.Status.of(status, statusMessage, metadata)));
+                        listeners.get("end").forEach(
+                                (item, index, arr) -> item.call(null,
+                                        ResponseStreamWrapper.Status.of(status, statusMessage, metadata)));
+                        listeners.clear();
+                    });
+                    props.setOnHeaders(headers -> {
+                        listeners.get("headers").forEach((item, index, arr) -> item.call(null, headers));
+                    });
+                    Request client = Grpc.invoke.onInvoke(BrowserFlightService.OpenHandshake, props);
+
+                    return new ResponseStream<HandshakeResponse>() {
+                        @Override
+                        public void cancel() {
+                            listeners.clear();
+                            client.getClose().onInvoke();
+                        }
+
+                        @Override
+                        public ResponseStream<HandshakeResponse> on(String type, Function handler) {
+                            listeners.get(type).push(handler);
+                            return this;
+                        }
+                    };
+                },
+                (next, headers, callback) -> connection.browserFlightServiceClient().nextHandshake(next, headers,
+                        callback::apply),
+                new HandshakeRequest());
+    }
+
+    private static Map<String, JsArray<Function>> listenerMap() {
+        Map<String, JsArray<Function>> listeners = new HashMap<>();
+        listeners.put("data", new JsArray<>());
+        listeners.put("end", new JsArray<>());
+        listeners.put("status", new JsArray<>());
+        listeners.put("headers", new JsArray<>());
+        return listeners;
```

Comment:
These constants for the type is used a few times in the class, should probably be extracted to a constant

Reply:
And then rewritten to never do this crap again. But yes.